request.go

```diff
@@ -1,29 +1,90 @@
 package sarama
 
-type requestEncoder interface {
+import (
+	"fmt"
+)
+
+type requestHeader struct {
+	length        int32
+	key           int16
+	version       int16
+	correlationID int32
+	clientID      string
+}
+
+func (r *requestHeader) decode(pd packetDecoder) (err error) {
+	if r.length, err = pd.getInt32(); err != nil {
+		return err
+	}
+	if r.key, err = pd.getInt16(); err != nil {
+		return err
+	}
+	if r.version, err = pd.getInt16(); err != nil {
+		return err
+	}
+	if r.correlationID, err = pd.getInt32(); err != nil {
+		return err
+	}
+	r.clientID, err = pd.getString()
+	return nil
+}
+
+type request interface {
 	encoder
+	decoder
 	key() int16
 	version() int16
 }
 
-type request struct {
-	correlationID int32
-	id            string
-	body          requestEncoder
+func requestEncoder(req request, correlationID int32, clientID string) encodeFunc {
```

Comment:
ya, this is actually turning into a bigger refactor in my head, let me play with this a bit